tests/unit/test_index.py

```diff
@@ -165,32 +166,59 @@ def test_get_formatted_locations_basic_auth():
 
 
 @pytest.mark.parametrize(
-    ("egg_info", "search_name", "expected"),
+    ("egg_info", "canonical_name", "expected"),
+    [
+        # Trivial.
+        ("pip-18.0", "pip", 3),
+        ("zope-interface-4.5.0", "zope-interface", 14),
+
+        # Canonicalized name match non-canonicalized egg info. (pypa/pip#5870)
+        ("Jinja2-2.10", "jinja2", 6),
+        ("zope.interface-4.5.0", "zope-interface", 14),
+        ("zope_interface-4.5.0", "zope-interface", 14),
+
+        # Should be smart enough to parse ambiguous names from the provided
+        # package name.
+        ("foo-2-2", "foo", 3),
+        ("foo-2-2", "foo-2", 5),
+
+        # Should be able to detect collapsed characters in the egg info.
+        ("foo--bar-1.0", "foo-bar", 8),
+        ("foo-_bar-1.0", "foo-bar", 8),
+    ],
+)
+def test_find_name_version_sep(egg_info, canonical_name, expected):
```

Comment:
Do we know in advance that all of the `egg_info` strings passed to the function will be valid? Or is there the possibility of mistaken input by the user? If the latter, I think it would be worth having some test cases to make sure e.g. the function doesn't treat invalid input as valid. (More generally, what guarantees do we have about what will be passed to the function?)

Reply:
The egg info strings come from an index or find links page. They can be invalid only if the page is formatted incorrectly. I think it is reasonable to either expect invalid inputs or not, depending on how you think pip should handle non-standard-conforming inputs.

Reply:
Okay, what I was getting at is that I think it would be worth adding a few more test cases that came to mind, even if invalid -- not because the additional cases should be considered valid, but because it provides more visibility into how the two functions behave on various edge cases. If it's possible for these inputs to happen in real life, then that's even more reason.

For example, I noticed that none of the `_find_name_version_sep()` test cases hit the `ValueError` line (though some of the `_egg_info_matches()` test cases hit that code path by virtue of calling `_find_name_version_sep()`).

Here are the additional test cases that I was thinking of (for both functions):

```
("zope.interface.-4.5.0", "zope-interface"),
("zope.interface--4.5.0", "zope-interface"),
("zope.interface--", "zope-interface"),
("zope.interface-", "zope-interface"),
("zope.interface", "zope-interface"),
```

(You may want to put the test cases for `_find_name_version_sep()` that cause a `ValueError` into its own test function.)